lib/optimist.rb

```diff
@@ -796,32 +799,80 @@ def full_description
     desc_str
   end
 
+  ## Format stdio like objects to a string
+  def format_stdio(obj)
+    case obj
+    when $stdout   then '<stdout>'
+    when $stdin    then '<stdin>'
+    when $stderr   then '<stderr>'
+    else obj # pass-through-case
+    end
+  end
+
   ## Generate the default value string for the educate line
   private def default_description_str str
     default_s = case default
-                when $stdout   then '<stdout>'
-                when $stdin    then '<stdin>'
-                when $stderr   then '<stderr>'
                 when Array
                   default.join(', ')
                 else
-                  default.to_s
+                  format_stdio(default).to_s
                 end
     defword = str.end_with?('.') ? 'Default' : 'default'
     " (#{defword}: #{default_s})"
   end
 
+  def permitted_valid_string
+    case permitted
+    when Array
+      return "one of: " + permitted.to_a.map(&:to_s).join(', ')
+    when Range
+      return "value in range of: #{permitted}"
+    when Regexp
+      return "value matching: #{permitted.inspect}"
+    end
+    raise NotImplementedError, "invalid branch"
+  end
+
+  def permitted_type_valid?
+    case permitted
+    when NilClass, Array, Range, Regexp then true
+    else false
+    end
+  end
+
+  def validate_permitted(arg, value)
+    return true if permitted.nil?
+    unless permitted_value?(value)
+      format_hash = {arg: arg, given: value, value: value, valid_string: permitted_valid_string(), permitted: permitted }
+      raise CommandlineError, permitted_response % format_hash
+    end
+    true
+  end
+  
+  # incoming values from the command-line should be strings, so we should
+  # stringify any permitted types as the basis of comparison.
+  def permitted_value?(val)
+    case permitted
+    when nil then true
+    when Regexp then val.match permitted
+    when Range then permitted.to_a.map(&:to_s).include? val
```

Comment:
iirc I did this b/c of the typing mismatch between the given arg from the cmdline (always a string) and the range type, though maybe it could be compared the other way around?

Reply:
Offhand, the way you did it is the only way I can think of, so overall I'm fine to merge.  We get strings from the command line so there's really not much we can do about that.